### PR TITLE
Feature/114 publish key kafka

### LINKED ISSUE #114: Publish key from kafka source as key to kafka sink / HyperdriveContext
Currently, the key from a kafka source is not ingested in `ConfluentAvroKafkaStreamDecoder`. Also, no key is produced to the kafka sink in `KafkaStreamWriter`. It should be possible to publish the ingested keys along with the value.

Consuming keys is supported by Abris like so
```
val result: DataFrame  = dataFrame.select(
    from_confluent_avro(col("key"), keyRegistryConfig) as 'key,
    from_confluent_avro(col("value"), valueRegistryConfig) as 'value)
```

**Tasks**
- Add config properties for decoder
    - `decoder.avro.consume.keys` (true / false)
    - `decoder.avro.key.schema.naming.strategy`
    - `decoder.avro.key.schema.id`
    - `decoder.avro.key.schema.record.name`
    - `decoder.avro.key.schema.record.namespace`

- Add config properties for writer
    - `writer.kafka.produce.keys` (true / false)
    - `writer.kafka.key.schema.naming.strategy`
    - `writer.kafka.key.schema.record.name`
    - `writer.kafka.key.schema.record.namespace`

- Add context object (`HyperdriveContext`) to which a value can be put and retrieved by key
    - To avoid key collisions, used keys should be documented
    - `ConfluentAvroKafkaStreamDecoder` should put the key column names, `KafkaStreamWriter` should retrieve them

### PR DESCRIPTION
**Pull request by kevinwallimann**
Closes #114

### CODE REVIEW

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+package za.co.absa.hyperdrive.ingestor.api.context
+
+import scala.collection.{mutable => mu}
+import scala.reflect.ClassTag
+import scala.util.{Failure, Success, Try}
+import scala.reflect.runtime.{universe => ru}
+
+object HyperdriveContext {
+  private val contextValues = mu.Map[String, Any]()
```
Comment (kevinwallimann) [inline api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:24]: Immutable var should be preferred. https://stackoverflow.com/questions/11386559/val-mutable-versus-var-immutable-in-scala

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+  def put(key: String, value: Any): Option[Any] = contextValues.put(key, value)
+
+  def get[T:ClassTag:ru.TypeTag](key: String): Try[T] = {
```
Comment (kevinwallimann) [inline api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:28]: Is ClassTag required? Maybe TypeTag is a superset for ClassTag

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    val value: Try[Any] = contextValues.get(key)
+      .map(Success(_))
+      .getOrElse(Failure(new NoSuchElementException(s"Could not find value for key $key")))
+
```
Comment (kevinwallimann) [inline api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:32]: Probably this can be simplified with recovers

File: ingestor-default/src/test/scala/za/co/absa/hyperdrive/ingestor/implementation/utils/TestSchemaRegistrySettingsUtil.scala [hunk 4]
```diff
@@ -15,17 +15,183 @@
...
 class TestSchemaRegistrySettingsUtil extends FlatSpec with Matchers {
 
   behavior of SchemaRegistrySettingsUtil.getClass.getName
 
-  "namingStrategyInvolvesRecord" should "return true only for record strategies" in {
-    SchemaRegistrySettingsUtil.namingStrategyInvolvesRecord(RECORD_NAME) shouldBe true
-    SchemaRegistrySettingsUtil.namingStrategyInvolvesRecord(TOPIC_RECORD_NAME) shouldBe true
-    SchemaRegistrySettingsUtil.namingStrategyInvolvesRecord(TOPIC_NAME) shouldBe false
-    SchemaRegistrySettingsUtil.namingStrategyInvolvesRecord("anything") shouldBe false
+  private val topic = "topic"
```
Comment (kevinwallimann) [inline ingestor-default/src/test/scala/za/co/absa/hyperdrive/ingestor/implementation/utils/TestSchemaRegistrySettingsUtil.scala:27]: Move vals to companion object

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala (hunk 1 above)
Comment (yruslan) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:24]: I think the current implementation is good as well since the map never escapes the scope and there is no intention to expose the map in the future.

File: api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/context/TestHyperdriveContext.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.ingestor.api.context
+
+import org.scalatest.{FlatSpec, Matchers}
+
+class TestHyperdriveContext extends FlatSpec with Matchers {
```
Comment (jozefbakus) [inline api/src/test/scala/za/co/absa/hyperdrive/ingestor/api/context/TestHyperdriveContext.scala:20]: I would add one test for duplicate key

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala (hunk 2 above)
Comment (kevinwallimann) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:28]: @benedeki I've just tested: The ClassTag is indeed required. And the TypeTag is only required for the `ru.typeOf[T]` in the error message

Comment (benedeki) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:28]: Thanks for testing. Interesting, I still need to study this (area) for better understanding.

File: api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala [hunk 6]
```diff
@@ -15,24 +15,23 @@
...
+  def put(key: String, value: Any): Unit = {
+    contextValues = contextValues + (key -> value)
+  }
 
-    value.flatMap {
-      case asT: T => Success(asT)
-      case _ => Failure(new ClassCastException(s"Could not cast value $value for key $key to type ${ru.typeOf[T]}"))
-    }
+  def get[T: ClassTag : ru.TypeTag](key: String): Try[T] = {
+    Try(
```
Comment (benedeki) [inline api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:30]: This is a lovely piece of code 👍

Comment (benedeki) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:30]: Hm, just out of curiosity, how does it behave, when the values is of type `Int` but requested `Long`?

Comment (kevinwallimann) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:30]: Thanks :). If you request an `Int` as a `Long` it throws the ClassCastException since you can't cast `Int` to `Long` in Scala with a pattern match.

Comment (jozefbakus) [reply api/src/main/scala/za/co/absa/hyperdrive/ingestor/api/context/HyperdriveContext.scala:30]: > This is a lovely piece of code 👍

I agree !
